File: ruby/luhn/luhn.rb
# Luhn
class Luhn
  attr_reader :num
  def initialize(num)
    @num = num
  end

  def addends
    num.to_s.chars.map(&:to_i) # digits
       .reverse
       .map.with_index { |digit, index| addend(digit, index) }
       .reverse
  end

  def addend(digit, index)
    subtract_nine(double_every_other(digit, index))
  end

  def checksum
    addends.inject(:+)
  end

  def valid?
    checksum % 10 == 0
  end

  def self.create(num)
    (0..9).each do |i|
      return "#{num}#{i}".to_i if Luhn.new("#{num}#{i}".to_i).valid?
    end
  end

  private

  def double_every_other(digit, index)
    index.odd? ? digit * 2 : digit
  end

  def subtract_nine(digit)
    digit > 9 ? digit - 9 : digit
  end
end

File: python/luhn/luhn.py
class Luhn:

    def __init__(self, number):
        self.number = number

    def checksum(self):
        return sum(self.addends()) % 10

    def addends(self):
        return [self.addend(idx, int(val)) for idx, val in
                enumerate(reversed(str(self.number)))]

    def addend(self, idx, val):
        return self.subtract_nine(idx, self.double_every_other(idx, val))

    def double_every_other(self, idx, val):
        return val * 2 if idx % 2 == 1 else val

    def subtract_nine(self, idx, val):
        return val - 9 if val > 10 else val

    def is_valid(self):
        return self.checksum() == 0

    @classmethod
    def create(cls, num):
        for i in range(0, 10):
            if cls(int(str(num) + str(i))).is_valid():
                return int(str(num) + str(i))
